### PR TITLE
feat: introduce decider-agent

### PR DESCRIPTION
**Pull request by fengsh27**

### CODE REVIEW

**Comment by slobentanzer** [comment]
Thanks, looking good. Specifically about the retries, I think it's good to handle this on a per-tool basis. E.g., BLAST can take a while until computation completes, that's why a couple more retries are needed until the result can be fetched. I would increase the number for those tools, but 3 is a good default for APIs that should immediately return something.

File: biochatter/database_agent.py [hunk 1]
```diff
@@ -107,3 +111,24 @@ def get_query_results(self, query: str, k: int = 3) -> list[Document]:
...
+        try:
+            result = self.driver.query("MATCH (n:Schema_info) RETURN n LIMIT 1")
+
+            if result[0]:
+                schema_info_node = result[0][0]["n"]
+                MAX_SCHEMA_INFO_LENGTH = 1000
+                schema_dict_content = schema_info_node["schema_info"][:MAX_SCHEMA_INFO_LENGTH] # limit to 1000 characters
+                return (f"the graph database contains the following nodes and edges: \n\n"
+                        f"{schema_dict_content}")
+        except Exception:
```
Comment (slobentanzer) [inline biochatter/database_agent.py:125]: @fengsh27 can we be more specific here? what kind of exception do we expect? do we need to inform the user this happened (logging, printing, raising)?

File: biochatter/vectorstore_agent.py [hunk 2]
```diff
@@ -625,3 +625,19 @@ def get_all_documents(
...
+                    return meta[col]
+            return ""
+
+        expr = VectorDatabaseAgentMilvus._build_meta_col_query_expr_for_all_documents(doc_ids)
+        result = self._col_metadata.query(
+            expr=expr, output_fields=METADATA_FIELDS,
+        )
+        names = list(map(get_name, result))
+        names_set = set(names)
+        return f"This vector store contains the following articles: {names_set}"
```
Comment (slobentanzer) [inline biochatter/vectorstore_agent.py:643]: Do we need to limit this for large vector stores? Should we give an option to manually describe a vector store, going beyond the names of articles in it?

Comment (fengsh27) [reply biochatter/vectorstore_agent.py:646]: It's a good point, I think we can enable to add descriptions in the customization yaml file and pass them to RagAgent

File: biochatter/database_agent.py (hunk 1 above)
Comment (fengsh27) [reply biochatter/database_agent.py]: it's not appropriate to use try-catch here

File: biochatter/vectorstore_agent.py (hunk 2 above)
Comment (fengsh27) [reply biochatter/vectorstore_agent.py]: @slobentanzer  I added RagAgent.agent_desc, which will override description of DatabaseAgent, VectorstroeAgent and API agent if it is set

**Comment by slobentanzer** [comment]
@fengsh27 can you try again?

**Comment by fengsh27** [comment]
@slobentanzer No it seems it doesn't work

**Comment by slobentanzer** [comment]
Is it the `ci.yaml` in this repository? You should have write access...

Do you have a more detailed error description?

**Comment by fengsh27** [comment]
You should be able to see the error information:
image

**Comment by fengsh27** [comment]
Here is error message:
```
Switched to branch 'coverage'

##[info] Updating branch coverage
##[info] Removing all files from target branch
##[info] Copying all files from /home/runner/work/biochatter/biochatter/docs/coverage
##[info] Pushing
remote: Permission to biocypher/biochatter.git denied to github-actions[bot].
fatal: unable to access 'https://github.com/biocypher/biochatter.git/': The requested URL returned error: 403

Error: Process exited with code: 1[28](https://github.com/biocypher/biochatter/actions/runs/10150872160/job/28068937469?pr=183#step:8:29):
remote: Permission to biocypher/biochatter.git denied to github-actions[bot].
fatal: unable to access 'https://github.com/biocypher/biochatter.git/': The requested URL returned error: 403

    at ChildProcess.<anonymous> (/home/runner/work/_actions/s0/git-publish-subdir-action/develop/action/dist/index.js:11966:20)
    at ChildProcess.emit (node:events:519:28)
    at maybeClose (node:internal/child_process:1105:16)
    at ChildProcess._handle.onexit (node:internal/child_process:[30](https://github.com/biocypher/biochatter/actions/runs/10150872160/job/28068937469?pr=183#step:8:31)5:5)
```

**Comment by fengsh27** [comment]
@slobentanzer I think I don't have permission to repo biocypher/biochatter. I forked the repo, and can only submit the changes by raising PR.

**Comment by slobentanzer** [comment]
Ah, these are not your permissions, but the GitHub bot. I recently introduced coverage to the CI, but it does not seem to work on PRs (this is the first PR that has the change I think). This is the changed CI:

```
      - name: Run tests and generate coverage report
        run: poetry run coverage run -m pytest test --ignore=./volumes
        shell: bash

      - name: Generate coverage badge
        run: poetry run coverage-badge -f -o docs/coverage/coverage.svg
        shell: bash

      - name: Commit changes
        uses: s0/git-publish-subdir-action@develop
        env:
          REPO: self
          BRANCH: coverage
          FOLDER: docs/coverage
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

I guess we cannot update the `coverage` branch from a PR (nor do we want to). Maybe this can be fixed by simply running the last step only on `main`, not on PRs. Let me try.

**Comment by slobentanzer** [comment]
@fengsh27 now it works :)

**Comment by fengsh27** [comment]
@slobentanzer is this PR ready to merge?

**Comment by slobentanzer** [comment]
@fengsh27 have not had time for thorough review, is next on my list. Looking into it now (after dealing with some issues on the biochatter-server Docker).

**Review by slobentanzer — APPROVED** [review]
Great job! Merging and incrementing minor to `0.6.0` due to the significant additions.

**Pull request merged**